timesketch/lib/google_auth.py

```diff
@@ -34,6 +47,135 @@ class JwtKeyError(Exception):
+    return resp.json()
+
+
+def _fetch_oauth2_discovery_document():
+    """Fetch Google OAuth2 discovery document.
+
+    Raises:
+        DiscoveryDocumentError if document cannot be fetched.
+
+    Returns:
+        HTTP response.
+    """
+    try:
+        resp = requests.get(DISCOVERY_URL)
+    except requests.exceptions.RequestException as e:
+        raise DiscoveryDocumentError(
+            'Cannot fetch discovery document: {}'.format(e))
+    if resp.status_code != HTTP_STATUS_CODE_OK:
+        raise DiscoveryDocumentError(
+            'Cannot fetch discovery_document: {}'.format(resp.status_code))
+    return resp.json()
+
+
+def _generate_random_token():
+    """Generate random string to use as CSRF and nonce tokens.
+
+    Returns:
+        Random string.
+    """
+    return hashlib.sha256(os.urandom(1024)).hexdigest()
+
+
+def get_oauth2_authorize_url(hosted_domain=None):
+    """Generate an authorization URL for Google's OAuth2 service.
+
+    Args:
+        hosted_domain: Optional GSuite domain to limit access to.
+
+    Returns:
+        Authorization URL.
+    """
+    csrf_token = _generate_random_token()
+    nonce = _generate_random_token()
+    redirect_uri = url_for('user_views.google_openid_connect', _external=True)
+    scopes = ('openid', 'email', 'profile')
+
+    # Add the generated CSRF token to the client session for later validation.
+    session[CSRF_KEY] = csrf_token
+
+    # Generate authorization URL
+    params = dict(
+        client_id=current_app.config.get('GOOGLE_OIDC_CLIENT_ID'),
+        scope=' '.join(scopes),
+        response_type='code',
+        access_type='online',  # Because we don't need a refresh token.
+        state=csrf_token,
+        nonce=nonce,  # Enable replay attack protection attack.
+        redirect_uri=redirect_uri
+    )
+    if hosted_domain:
+        params['hd'] = hosted_domain
+
+    urlencoded_params = urllib.urlencode(params)
+    google_authorization_url = '{}?{}'.format(AUTH_URI, urlencoded_params)
+    return google_authorization_url
+
+
+def get_encoded_jwt_over_https(code):
+    """Fetch a JSON Web Token (JWT) using a authentication code.
+
+    Args:
+        code: Authentication code obtained from an OAuth2 flow.
+
+    Raises:
+        JwtFetchError if JWT cannot be fetched.
+
+    Returns:
+        Encoded JWT.
+    """
+
+    discovery_document = get_oauth2_discovery_document()
+    redirect_uri = url_for('user_views.google_openid_connect', _external=True)
+    post_data = {
+        'code': code,
+        'client_id': current_app.config.get('GOOGLE_OIDC_CLIENT_ID'),
+        'client_secret': current_app.config.get('GOOGLE_OIDC_CLIENT_SECRET'),
+        'redirect_uri': redirect_uri,
+        'grant_type': 'authorization_code'
+    }
+    token_url = discovery_document.get('token_endpoint')
+    try:
+        response = requests.post(token_url, data=post_data)
+        encoded_jwt = response.json().get('id_token')
```

Comment:
Maybe account for the situation where `id_token` is not in the response?

Reply:
Done.